lib/mutgen.c

```diff
@@ -25,66 +25,350 @@
 
 #include "msprime.h"
 
+
+typedef struct mutation_t {
+    tsk_id_t id;
+    tsk_id_t site;
+    tsk_id_t node;
+    tsk_id_t parent;
+    char *derived_state;
+    tsk_size_t derived_state_length;
+    char *metadata;
+    tsk_size_t metadata_length;
+    // additions to tsk_mutation_t:
+    double time;
+    struct mutation_t *prev;
+    struct mutation_t *next;
+    bool new;
+} mutation_t;
+
 typedef struct {
-    const char *ancestral_state;
-    const char *derived_state;
-} mutation_type_t;
+    double position;
+    char *ancestral_state;
+    tsk_size_t ancestral_state_length;
+    char *metadata;
+    tsk_size_t metadata_length;
+    // modifications to tsk_site_t:
+    mutation_t *mutations;
+    size_t mutations_length;
+    bool new;
+} site_t;
+
+static int
+cmp_site(const void *a, const void *b)
+{
+    const site_t *ia = (const site_t *) a;
+    const site_t *ib = (const site_t *) b;
+    return (ia->position > ib->position) - (ia->position < ib->position);
+}
+
+static int
+cmp_mutationp(const void *a, const void *b)
```

Comment:
I don't follow the comparison logic here either, but made that comment in Python.